src/client/PyCallbacks.h

```diff
@@ -97,25 +101,44 @@ class PyCallbacks : public IPyCallbacks {
 		return mModels[initialShapeIdx].mCGAReport;
 	}
 
+	const std::wstring& getCGAPrints(const size_t initialShapeIdx) const {
+		if (initialShapeIdx >= mModels.size())
+			throw std::out_of_range("initial shape index is out of range.");
+
+		return mModels[initialShapeIdx].mCGAPrints;
+	}
+
+	const std::wstring& getCGAErrors(const size_t initialShapeIdx) const {
+		if (initialShapeIdx >= mModels.size())
+			throw std::out_of_range("initial shape index is out of range.");
+
+		return mModels[initialShapeIdx].mCGAErrors;
+	}
+
 	prt::Status generateError(size_t isIndex, prt::Status status, const wchar_t* message) override {
-		pybind11::print(L"GENERATE ERROR:", isIndex, status, message);
 		return prt::STATUS_OK;
 	}
 
 	prt::Status assetError(size_t isIndex, prt::CGAErrorLevel level, const wchar_t* key, const wchar_t* uri,
 	                       const wchar_t* message) override {
-		pybind11::print(L"ASSET ERROR:", isIndex, level, key, uri, message);
+		std::wstring errorMsg(L"Asset" + ERRORLEVELS[level] + key + L" " + uri + L" " + message);
+		mModels[isIndex].mCGAErrors += errorMsg;

```

Comment:
We need to ensure that a newline is added before the new message is appended...

Reply:
I preferred creating a list of errors. In this case, there is no more need of an additional newline character.